automatically fix version in conf.py

aman-roy opened the pull request:
**Link the Issue(s) this Pull Request is related to.**
Fixes #591

**Summarize your change.**
 [docs/conf.py](https://github.com/AcademySoftwareFoundation/OpenCue/blob/master/docs/conf.py) has two variables, `version` and `release` which is as of now, statically typed. This script changes these two with the current `version` and `release`.

It's also helpful to describe **why** you're making this change.
Doing this manually was not a good solution at all. Having it as an automated process will make it more hassle-free.

sharifsalah commented:
Thanks @aman-roy, I've tested your code  and overall this looks good to me. It certainly does what we discussed in the issue.

I'll leave it to the code owners to review the pull request in detail.

One small suggestion would be to add a very brief comment to `conf.py` to explain to developers that they can run this script from the `docs` directory; something like:

```
# To set the following values to the latest version and 
# release, run `python3 docs/version_release_fix.py`
```

aman-roy commented:
@bcipriano 

If you check the discussion we had in issue #591, Everyone suggested that `git tag` is the best way to get the versions, because [ ci/generate_version_number.sh](https://github.com/AcademySoftwareFoundation/OpenCue/blob/master/ci/generate_version_number.sh) generates version like - 

> if run from the master branch: [version from VERSION.in].[number of commits]
> if run from any other branch: [version from VERSION.in].[number of commits in master branch +1]-[short hash of the commit]

And as you said earlier,

>  Last tagged is likely what we're looking for.

bcipriano commented:
Yes, I'm aware of the previous discussion on the issue.

Now that I see it in action, I'm reconsidering that approach. I think it's problematic, for the reasons stated in my review comment.

aman-roy commented:
Should I keep the version as it is, and do the changes for the release part? 
For release, I will use the VERSION file which will get generated by [ci/generate_version_number.sh](https://github.com/AcademySoftwareFoundation/OpenCue/blob/master/ci/generate_version_number.sh)

bcipriano commented:
Please update `version_release_fix.py` to assume the `VERSION` file exists and read that in for version information. If it doesn't exist, the script should fail gracefully, with an informative error message.

bcipriano commented:
And thank you for bearing with us on this. Sometimes the consequences of changes aren't clear until you start to do the work. :)

aman-roy commented:
No issue, I understand. I will fix it and I truly appreciate the suggestions you gave.

aman-roy commented:
There is only one thing I am confused about.

Two fields are there, version and release. Do they both rely on VERSION file which will get generated?

bcipriano commented:
I think they can use the same information, yes. Based on the comments in `conf.py`, `release` will be equal to the full version number e.g. `0.3.43` and `version` is just the first two components of that number e.g. `0.3`.

`VERSION` has the full number, so it can become `release` as-is and for `version` just drop everything after the last period.

aman-roy commented:
@bcipriano @sharifsalah I have made the changes according to what you both suggested.

- Added comment in `conf.py` for running `version_release_fix.py`  first for fixing the `version` and `release`.
- Removed dependency from` git tag` to `VERSION` file for fetching `version` and `release`.

bcipriano commented on docs/version_release_fix.py:
```diff
@@ -0,0 +1,37 @@
+
+# Update version
+new_file = re.sub(r"version = u'[0-9.]+'", \
+	f"version = u'{version}'", old_file)
```
Please use str.format instead of f-strings, which aren't available when running Python 2.

aman-roy commented:
@bcipriano 
I have refactored the code again. I think now it is ready to merged!

bcipriano commented on docs/version_release_fix.py:
```diff
@@ -0,0 +1,42 @@
+import re, os
+
+def main():
+	cur_dir = os.path.dirname(__file__)
```
nit: Please use two-space indents instead of tabs, this matches the style used in the rest of the code.

aman-roy replied:
fixed whitespace. please check.

bcipriano commented:
@smith1511 @gregdenton @larsbijl Need a second review here please!

Also @sharifsalah if you want another look at this that would be welcome.

sharifsalah commented:
I'll do my best to take another look tomorrow.

sharifsalah commented on docs/conf.py:
```diff
@@ -11,7 +11,11 @@
 # documentation root, use os.path.abspath to make it absolute, like shown here.
-#
+
+# To set the following values to the latest version and
```
I suggest you move this comment so that it sits just before (current) lines 31-34,  similar to the following:

```
# To set the following values to the latest version and
# release, run `python3 docs/version_release_fix.py`

# The short X.Y version
version = u'0.3.6'
# The full version, including alpha/beta/rc tags
release = u'0.3.6'
```

Otherwise it's might not be clear to the user what 'the following values' is referring to.

aman-roy replied:
I have made the changes. Please review.

sharifsalah replied:
LGTM, thanks!

sharifsalah commented:
Looks great, thanks for your work on this @aman-roy

aman-roy commented:
Thanks @bcipriano, @sharifsalah for all the helpful suggestions.
